### PR TITLE
API: Unembed Put structs from Get structs

### PR DESCRIPTION
**Pull request by MusicDin**
Unembed Put structs from Get structs.

Fixes https://github.com/canonical/lxd/issues/12993

### CODE REVIEW

File: lxd/patches.go [hunk 1]
```diff
@@ -325,7 +325,8 @@ func patchNetworkACLRemoveDefaults(name string, d *Daemon) error {
 
 				// Write back modified config if needed.
 				if modified {
-					err = tx.UpdateNetworkACL(ctx, aclID, &acl.NetworkACLPut)
+					aclPut := acl.Writable()
+					err = tx.UpdateNetworkACL(ctx, aclID, &aclPut)
```
Comment (MusicDin) [inline lxd/patches.go:329]: Is there any reason *UpdateX* functions accept pointers to Put structures? 
Is it possible that some of them modify contents of the provided structure? Because in such case the updated Put struct has to be written back to the Get struct.

File: shared/api/network_forward.go [hunk 2]
```diff
@@ -139,5 +158,9 @@ func (f *NetworkForward) Etag() []any {
 
 // Writable converts a full NetworkForward struct into a NetworkForwardPut struct (filters read-only fields).
 func (f *NetworkForward) Writable() NetworkForwardPut {
-	return f.NetworkForwardPut
+	return NetworkForwardPut{
+		Description: f.Description,
+		Config:      f.Config,
+		Ports:       f.Ports,
+	}
 }
```
Comment (MusicDin) [inline shared/api/network_forward.go:166]: Since Put struct cannot be directly set to the Get structures anymore, should we provide `SetWritable()` function that would allow doing that?

File: shared/api/certificate.go [hunk 3]
```diff
@@ -72,7 +72,31 @@ type CertificatePut struct {
...
+	// API extension: certificate_project
+	Restricted bool `json:"restricted" yaml:"restricted"`
+
+	// List of allowed projects (applies when restricted)
+	// Example: ["default", "foo", "bar"]
+	//
+	// API extension: certificate_project
+	Projects []string `json:"projects" yaml:"projects"`
+
+	// The certificate itself, as PEM encoded X509 (or as base64 encoded X509 on POST)
```
Comment (markylaing) [inline shared/api/certificate.go:95]: If we're unembedding these API types, now might be a good time to also unembed `CertificatePut` from `CertificatesPost` so that the base64 encoding of the certificate on POST can be better reflected in the API docs.

File: shared/api/network_forward.go (hunk 2 above)
Comment (tomponline) [reply shared/api/network_forward.go:166]: Sounds good

File: shared/api/certificate.go (hunk 3 above)
Comment (tomponline) [reply shared/api/certificate.go:95]: Sounds good

File: lxd/patches.go (hunk 1 above)
Comment (tomponline) [reply lxd/patches.go:329]: Yeah i think it would be clearer to not accept a pointer where these functions dont modify the struct.

File: shared/api/certificate.go (hunk 3 above)
Comment (MusicDin) [reply shared/api/certificate.go:95]: Very good point. Will address that.

Comment (tomponline) [reply shared/api/certificate.go:119]: @MusicDin this comment doesnt seem appropriate here now.

File: lxd/cluster/connect.go [hunk 4]
```diff
@@ -223,8 +223,12 @@ func SetupTrust(serverCert *shared.CertInfo, serverName string, targetAddress st
 	}
 
 	post := api.CertificatesPost{
-		CertificatePut: cert.CertificatePut,
```
Comment (tomponline) [inline lxd/cluster/connect.go:226]: Shouldn't this be using `post.SetWritable(cert.CertificatePut)`?

File: lxd/db/cluster/auth_groups.go [hunk 5]
```diff
@@ -131,9 +129,9 @@ func (g *AuthGroup) ToAPI(ctx context.Context, tx *sql.Tx) (*api.AuthGroup, erro
 // GetIdentitiesByAuthGroupID returns the identities that are members of the group with the given ID.
 func GetIdentitiesByAuthGroupID(ctx context.Context, tx *sql.Tx, groupID int) ([]Identity, error) {
 	stmt := `
-SELECT identities.id, identities.auth_method, identities.type, identities.identifier, identities.name, identities.metadata 
-FROM identities 
-JOIN identities_auth_groups ON identities.id = identities_auth_groups.identity_id 
+SELECT identities.id, identities.auth_method, identities.type, identities.identifier, identities.name, identities.metadata
```
Comment (tomponline) [inline lxd/db/cluster/auth_groups.go:132]: query changes shouldnt be included with an unrelated commit message.

File: lxd/db/network_acls.go [hunk 6]
```diff
@@ -237,7 +235,7 @@ func networkACLConfigAdd(tx *sql.Tx, id int64, config map[string]string) error {
 }
 
 // UpdateNetworkACL updates the Network ACL with the given ID.
-func (c *ClusterTx) UpdateNetworkACL(ctx context.Context, id int64, config *api.NetworkACLPut) error {
+func (c *ClusterTx) UpdateNetworkACL(ctx context.Context, id int64, config api.NetworkACLPut) error {
```
Comment (tomponline) [inline lxd/db/network_acls.go:238]: should be a separate commit

File: lxd/db/network_forwards.go [hunk 7]
```diff
@@ -88,7 +88,7 @@ func networkForwardConfigAdd(tx *sql.Tx, forwardID int64, config map[string]stri
 }
 
 // UpdateNetworkForward updates an existing Network Forward.
-func (c *ClusterTx) UpdateNetworkForward(ctx context.Context, networkID int64, forwardID int64, info *api.NetworkForwardPut) error {
+func (c *ClusterTx) UpdateNetworkForward(ctx context.Context, networkID int64, forwardID int64, info api.NetworkForwardPut) error {
```
Comment (tomponline) [inline lxd/db/network_forwards.go:91]: commit message should explain why

File: lxd/db/network_load_balancers.go [hunk 8]
```diff
@@ -95,7 +95,7 @@ func networkLoadBalancerConfigAdd(tx *sql.Tx, loadBalancerID int64, config map[s
 }
 
 // UpdateNetworkLoadBalancer updates an existing Network Load Balancer.
-func (c *ClusterTx) UpdateNetworkLoadBalancer(ctx context.Context, networkID int64, loadBalancerID int64, info *api.NetworkLoadBalancerPut) error {
+func (c *ClusterTx) UpdateNetworkLoadBalancer(ctx context.Context, networkID int64, loadBalancerID int64, info api.NetworkLoadBalancerPut) error {
```
Comment (tomponline) [inline lxd/db/network_load_balancers.go:98]: commit message should say why

File: lxd/db/network_peers.go [hunk 9]
```diff
@@ -350,7 +350,7 @@ func (c *ClusterTx) GetNetworkPeerNames(ctx context.Context, networkID int64) (m
 }
 
 // UpdateNetworkPeer updates an existing Network Peer.
-func (c *ClusterTx) UpdateNetworkPeer(ctx context.Context, networkID int64, peerID int64, info *api.NetworkPeerPut) error {
+func (c *ClusterTx) UpdateNetworkPeer(ctx context.Context, networkID int64, peerID int64, info api.NetworkPeerPut) error {
```
Comment (tomponline) [inline lxd/db/network_peers.go:353]: commit message should say why

File: lxd/db/storage_buckets.go [hunk 10]
```diff
@@ -352,7 +352,7 @@ func storageBucketPoolConfigAdd(tx *sql.Tx, bucketID int64, config map[string]st
 }
 
 // UpdateStoragePoolBucket updates an existing Storage Bucket.
-func (c *ClusterTx) UpdateStoragePoolBucket(ctx context.Context, poolID int64, bucketID int64, info *api.StorageBucketPut) error {
+func (c *ClusterTx) UpdateStoragePoolBucket(ctx context.Context, poolID int64, bucketID int64, info api.StorageBucketPut) error {
```
Comment (tomponline) [inline lxd/db/storage_buckets.go:355]: commit message should say why

File: lxd/api_internal_recover.go [hunk 11]
```diff
@@ -164,10 +164,11 @@ func internalRecoverScan(s *state.State, userPools []api.StoragePoolsPost, valid
 
 				// If pool doesn't exist in DB, initialise a temporary pool with the supplied info.
 				poolInfo := api.StoragePool{
-					Name:           p.Name,
-					Driver:         p.Driver,
-					StoragePoolPut: p.StoragePoolPut,
```
Comment (tomponline) [inline lxd/api_internal_recover.go:169]: We have a p.StoragePoolPut here so we should be using api.StoragePool.SetWritable() here to avoid duplication.

File: lxd/network/driver_ovn.go [hunk 12]
```diff
@@ -4717,8 +4717,10 @@ func (n *ovn) ForwardUpdate(listenAddress string, req api.NetworkForwardPut, cli
 		}
 
 		newForward := api.NetworkForward{
-			ListenAddress:     curForward.ListenAddress,
-			NetworkForwardPut: req,
+			ListenAddress: curForward.ListenAddress,
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:4720]: As req is a api.NetworkForwardPut we should be using api.NetworkForward.SetWritable() here to avoid duplication.

File: lxd/network/driver_ovn.go [hunk 13]
```diff
@@ -5072,8 +5074,11 @@ func (n *ovn) LoadBalancerUpdate(listenAddress string, req api.NetworkLoadBalanc
 		}
 
 		newLoadBalancer := api.NetworkLoadBalancer{
-			ListenAddress:          curLoadBalancer.ListenAddress,
-			NetworkLoadBalancerPut: req,
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:5076]: As req is NetworkLoadBalancerPut we should be using api.NetworkLoadBalancer.SetWritable() to avoid duplication.

File: lxd/network/driver_ovn.go [hunk 14]
```diff
@@ -5579,8 +5584,9 @@ func (n *ovn) PeerUpdate(peerName string, req api.NetworkPeerPut) error {
 	}
 
 	newPeer := api.NetworkPeer{
-		Name:           curPeer.Name,
-		NetworkPeerPut: req,
+		Name:        curPeer.Name,
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:5587]: As req is NetworkPeerPut we should be using api.NetworkPeer.SetWritable() to avoid duplication.

**Review by tomponline — CHANGES_REQUESTED** [review]
some changes needed thanks!

**Comment by tomponline** [comment]
@MusicDin please may you rebase on https://github.com/canonical/lxd/pull/13045 as i intend to merge that one before this and it contains some API struct changes.

File: lxd/db/cluster/auth_groups.go (hunk 5 above)
Comment (MusicDin) [reply lxd/db/cluster/auth_groups.go:132]: This was not intentional, I guess it auto removed trailing spaces. Will split into separate commits

File: lxd/cluster/connect.go (hunk 4 above)
Comment (MusicDin) [reply lxd/cluster/connect.go:226]: This is Post structure which does not have SetWritable() currently. 
And Post is expecting base64 encoded certificate, while Put is not.
Wouldn't it get confusing if we add SetWritable() function?

Comment (tomponline) [reply lxd/cluster/connect.go:226]: ah ok thanks

**Comment by tomponline** [comment]
@MusicDin ive merged https://github.com/canonical/lxd/pull/13045 so can be rebased now.

File: shared/api/certificate.go (hunk 3 above)
Comment (MusicDin) [reply shared/api/certificate.go]: Thanks. Fixed.

File: lxd/db/cluster/identities.go [hunk 15]
```diff
@@ -306,18 +306,16 @@ func (i *Identity) ToAPI(ctx context.Context, tx *sql.Tx) (*api.Identity, error)
...
 	}, nil
 }
 
 // GetAuthGroupsByIdentityID returns a slice of groups that the identity with the given ID is a member of.
 func GetAuthGroupsByIdentityID(ctx context.Context, tx *sql.Tx, identityID int) ([]AuthGroup, error) {
 	stmt := `
-SELECT auth_groups.id, auth_groups.name, auth_groups.description 
-FROM auth_groups 
-JOIN identities_auth_groups ON auth_groups.id = identities_auth_groups.auth_group_id 
+SELECT auth_groups.id, auth_groups.name, auth_groups.description
```
Comment (tomponline) [inline lxd/db/cluster/identities.go:316]: oops

File: lxd/db/cluster/identities.go [hunk 16]
```diff
@@ -344,8 +342,8 @@ WHERE identities_auth_groups.identity_id = ?`
 // GetAllAuthGroupsByIdentityIDs returns a map of identity ID to slice of groups the identity with that ID is a member of.
 func GetAllAuthGroupsByIdentityIDs(ctx context.Context, tx *sql.Tx) (map[int][]AuthGroup, error) {
 	stmt := `
-SELECT identities_auth_groups.identity_id, auth_groups.id, auth_groups.name, auth_groups.description 
-FROM auth_groups 
+SELECT identities_auth_groups.identity_id, auth_groups.id, auth_groups.name, auth_groups.description
```
Comment (tomponline) [inline lxd/db/cluster/identities.go:345]: oops

File: lxd/db/cluster/identity_provider_groups.go [hunk 17]
```diff
@@ -66,9 +66,9 @@ func (i *IdentityProviderGroup) ToAPI(ctx context.Context, tx *sql.Tx) (*api.Ide
 // GetAuthGroupsByIdentityProviderGroupID returns a list of a groups that the identity provider group with the given ID.
 func GetAuthGroupsByIdentityProviderGroupID(ctx context.Context, tx *sql.Tx, idpGroupID int) ([]AuthGroup, error) {
 	stmt := `
-SELECT auth_groups.id, auth_groups.name, auth_groups.description 
-FROM auth_groups_identity_provider_groups 
-JOIN auth_groups ON auth_groups_identity_provider_groups.auth_group_id = auth_groups.id 
+SELECT auth_groups.id, auth_groups.name, auth_groups.description
```
Comment (tomponline) [inline lxd/db/cluster/identity_provider_groups.go:69]: oops

**Review by tomponline — CHANGES_REQUESTED** [review]
Looks good, just a few places that are removing white space from queries unrelated to commit message.

CC @markylaing please can you configure your editor to strip whitespace from end of lines as these seem to be predominately in the new auth queries.

**Comment by MusicDin** [comment]
> Looks good, just a few places that are removing white space from queries unrelated to commit message.

Fixed.

**Review by tomponline — DISMISSED** [review]
Thanks!

**Comment by MusicDin** [comment]
Seems I've broke something.. Looking into it.

**Comment by tomponline** [comment]
@MusicDin it looks fixed now, what was the issue? Thanks

**Comment by MusicDin** [comment]
Yep. I accidentally unset `Driver` field in StoragePool struct (in `lxd/api_internal_recover`) when changing to SetWritable().

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
Thanks!
